File: ios/RoutineChart/RoutineChart/Core/Data/Remote/Firebase/FirebaseAuthService.swift
//
//  FirebaseAuthService.swift
//  RoutineChart
//
//  Created for Phase 2.1: Firebase Auth
//

import Foundation
import FirebaseAuth
import Combine

/// Firebase implementation of AuthRepository
final class FirebaseAuthService: AuthRepository {
    private let auth = Auth.auth()
    private let authStateSubject = CurrentValueSubject<AuthUser?, Never>(nil)
    
    var currentUser: AuthUser? {
        guard let firebaseUser = auth.currentUser else { return nil }
        return AuthUser(
            id: firebaseUser.uid,
            email: firebaseUser.email,
            isAnonymous: firebaseUser.isAnonymous
        )
    }
    
    var authStatePublisher: AnyPublisher<AuthUser?, Never> {
        authStateSubject.eraseToAnyPublisher()
    }
    
    init() {
        // Initialize with current auth state
        authStateSubject.send(currentUser)
        
        // Listen for auth state changes
        auth.addStateDidChangeListener { [weak self] _, user in
            guard let self = self else { return }
            if let user = user {
                self.authStateSubject.send(AuthUser(
                    id: user.uid,
                    email: user.email,
                    isAnonymous: user.isAnonymous
                ))
            } else {
                self.authStateSubject.send(nil)
            }
        }
    }
    
    func signInWithEmail(email: String, password: String) async throws -> AuthUser {
        let result = try await auth.signIn(withEmail: email, password: password)
        return AuthUser(
            id: result.user.uid,
            email: result.user.email,
            isAnonymous: result.user.isAnonymous
        )
    }
    
    func signUpWithEmail(email: String, password: String) async throws -> AuthUser {
        let result = try await auth.createUser(withEmail: email, password: password)
        return AuthUser(
            id: result.user.uid,
            email: result.user.email,
            isAnonymous: result.user.isAnonymous
        )
    }
    
    func signInAnonymously() async throws -> AuthUser {
        let result = try await auth.signInAnonymously()
        return AuthUser(
            id: result.user.uid,
            email: result.user.email,
            isAnonymous: result.user.isAnonymous
        )
    }
    
    func linkAnonymousToEmail(email: String, password: String) async throws -> AuthUser {
        guard let currentUser = auth.currentUser, currentUser.isAnonymous else {
            throw AuthError.notAnonymous
        }
        
        let credential = EmailAuthProvider.credential(withEmail: email, password: password)
        let result = try await currentUser.link(with: credential)
        
        return AuthUser(
            id: result.user.uid,
            email: result.user.email,
            isAnonymous: result.user.isAnonymous
        )
    }
    
    func signOut() throws {
        try auth.signOut()
    }
    
    func sendPasswordReset(email: String) async throws {
        try await auth.sendPasswordReset(withEmail: email)
    }
}

// MARK: - Errors
enum AuthError: LocalizedError {
    case notAnonymous
    case invalidCredentials
    case userNotFound
    
    var errorDescription: String? {
        switch self {
        case .notAnonymous:
            return "Current user is not anonymous"
        case .invalidCredentials:
            return "Invalid email or password"
        case .userNotFound:
            return "User not found"
        }
    }
}

File: android/app/src/main/java/com/HammersTech/RoutineChart/core/data/remote/firebase/FirebaseAuthService.kt
package com.HammersTech.RoutineChart.core.data.remote.firebase

import com.HammersTech.RoutineChart.core.domain.models.AuthUser
import com.HammersTech.RoutineChart.core.domain.repositories.AuthRepository
import com.HammersTech.RoutineChart.core.utils.AppLogger
import com.google.firebase.auth.EmailAuthProvider
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.tasks.await
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Firebase implementation of AuthRepository
 * Phase 2.1: Firebase Auth
 */
@Singleton
class FirebaseAuthService
    @Inject
    constructor() : AuthRepository {
        private val auth = FirebaseAuth.getInstance()

        override val authStateFlow: Flow<AuthUser?> =
            callbackFlow {
                val listener =
                    FirebaseAuth.AuthStateListener { firebaseAuth ->
                        val user = firebaseAuth.currentUser?.toAuthUser()
                        trySend(user)
                    }

                auth.addAuthStateListener(listener)

                // Send initial value
                trySend(auth.currentUser?.toAuthUser())

                awaitClose {
                    auth.removeAuthStateListener(listener)
                }
            }

        override val currentUser: AuthUser?
            get() = auth.currentUser?.toAuthUser()

        override suspend fun signInWithEmail(
            email: String,
            password: String,
        ): Result<AuthUser> {
            return try {
                val result = auth.signInWithEmailAndPassword(email, password).await()
                val user = result.user?.toAuthUser()
                if (user != null) {
                    AppLogger.Auth.info("User signed in with email: $email")
                    Result.success(user)
                } else {
                    Result.failure(Exception("Failed to sign in"))
                }
            } catch (e: Exception) {
                AppLogger.Auth.error("Sign in failed", e)
                Result.failure(e)
            }
        }

        override suspend fun signUpWithEmail(
            email: String,
            password: String,
        ): Result<AuthUser> {
            return try {
                val result = auth.createUserWithEmailAndPassword(email, password).await()
                val user = result.user?.toAuthUser()
                if (user != null) {
                    AppLogger.Auth.info("User signed up with email: $email")
                    Result.success(user)
                } else {
                    Result.failure(Exception("Failed to create account"))
                }
            } catch (e: Exception) {
                AppLogger.Auth.error("Sign up failed", e)
                Result.failure(e)
            }
        }

        override suspend fun signInAnonymously(): Result<AuthUser> {
            return try {
                val result = auth.signInAnonymously().await()
                val user = result.user?.toAuthUser()
                if (user != null) {
                    AppLogger.Auth.info("User signed in anonymously")
                    Result.success(user)
                } else {
                    Result.failure(Exception("Failed to sign in anonymously"))
                }
            } catch (e: Exception) {
                AppLogger.Auth.error("Anonymous sign in failed", e)
                Result.failure(e)
            }
        }

        override suspend fun linkAnonymousToEmail(
            email: String,
            password: String,
        ): Result<AuthUser> {
            return try {
                val currentUser = auth.currentUser
                if (currentUser == null || !currentUser.isAnonymous) {
                    return Result.failure(Exception("Current user is not anonymous"))
                }

                val credential = EmailAuthProvider.getCredential(email, password)
                val result = currentUser.linkWithCredential(credential).await()
                val user = result.user?.toAuthUser()
                if (user != null) {
                    AppLogger.Auth.info("Anonymous account linked to email: $email")
                    Result.success(user)
                } else {
                    Result.failure(Exception("Failed to link account"))
                }
            } catch (e: Exception) {
                AppLogger.Auth.error("Account linking failed", e)
                Result.failure(e)
            }
        }

        override suspend fun signOut(): Result<Unit> {
            return try {
                auth.signOut()
                AppLogger.Auth.info("User signed out")
                Result.success(Unit)
            } catch (e: Exception) {
                AppLogger.Auth.error("Sign out failed", e)
                Result.failure(e)
            }
        }

        override suspend fun sendPasswordReset(email: String): Result<Unit> {
            return try {
                auth.sendPasswordResetEmail(email).await()
                AppLogger.Auth.info("Password reset email sent to: $email")
                Result.success(Unit)
            } catch (e: Exception) {
                AppLogger.Auth.error("Password reset failed", e)
                Result.failure(e)
            }
        }

        private fun FirebaseUser.toAuthUser(): AuthUser {
            return AuthUser(
                id = uid,
                email = email,
                isAnonymous = isAnonymous,
            )
        }
    }
